src/App.js

```diff
@@ -42,52 +41,78 @@ const NoLogin = () => {
   )
 }
 
-const App = ({ vms, visibility, config }) => {
-  let detailToRender = null
-  switch (visibility.get('dialogToShow')) {
-    case 'Options':
-      detailToRender = (<Options />)
-      break
+/**
+ * Main App component. Wrap the main react-router components together with
+ * the various dialogs and error messages that may be needed.
+ */
+class App extends React.Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      navigationConfirm: {
+        show: false,
+        confirmYes: undefined,
+        confirmNo: undefined,
+      },
+    }
+  }
+
+  resetUserConfirmation (show = false, confirmYes = undefined, confirmNo = undefined) {
+    this.setState({
+      navigationConfirm: { show, confirmYes, confirmNo },
```

Comment:
This is a really bad idea to store functions in component state, because `State it's a serializable* representation of one point in time—a snapshot.`
https://github.com/uberVU/react-guide/blob/master/props-vs-state.md

Reply:
What's a better way?

Reply:
Define those functions like methods.